File: src/main/java/com/morak/performancetracker/aop/rest/RestAdvice.java
package com.morak.performancetracker.aop.rest;

import com.morak.performancetracker.context.Accumulator;
import org.aopalliance.intercept.MethodInterceptor;
import org.aopalliance.intercept.MethodInvocation;

public class RestAdvice implements MethodInterceptor {

    private final RestMonitor restMonitor;
    private final Accumulator accumulator;

    public RestAdvice(RestMonitor restMonitor, Accumulator accumulator) {
        this.restMonitor = restMonitor;
        this.accumulator = accumulator;
    }

    @Override
    public Object invoke(MethodInvocation invocation) throws Throwable {
        Object argument = invocation.getArguments()[0];
        restMonitor.start(argument.toString());
        Object returnValue = invocation.proceed();
        restMonitor.end();
        accumulator.add(restMonitor);
        return returnValue;
    }
}

File: src/main/kotlin/com/morak/performancetracker/aop/rest/RestAdvice.kt
package com.morak.performancetracker.aop.rest

import com.morak.performancetracker.context.Accumulator
import org.aopalliance.intercept.MethodInterceptor
import org.aopalliance.intercept.MethodInvocation

class RestAdvice(private val restMonitor: RestMonitor, private val accumulator: Accumulator) : MethodInterceptor {
    override fun invoke(invocation: MethodInvocation): Any? {
        val argument = invocation.arguments[0]
        restMonitor.start(argument.toString())
        val returnValue = invocation.proceed()
        restMonitor.end()
        accumulator.add(restMonitor)
        return returnValue
    }
}
